### DISCUSSION TITLE
When I use Shell script to run PyTest, some ERROR jumped up.

### DISCUSSION THREAD

**Discussion by lucky33newman** [pytest-dev/pytest#10539] [Q&A] [answered]
Hello,

When I use Shell script to run pytest. In the Shell script, the contents are:

```bash
#!/usr/bin/sh

cd /usr/pytest/
pytest -m "not dangerous"
```

It works well, but when I write the Shell script like this:

```bash
#!/usr/bin/sh

suffix="-m 'not dangerous'"

cd /usr/pytest/

pytest ${suffix}
```

I just get a ERROR message like this:

```
collected 0 items

========== no tests ran in 1.24s =================
ERROR: file or directory not found: dangerous'
```

I mean to run command like [pytest -m 'not dangerous']

Why it became like this?

I just can not understand.

Help!

**Comment by The-Compiler** [comment, marked as answer]
Please format code [as code](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) in the future, as your post is almost unreadable. I edited it to add code blocks.

This isn't a pytest issue, it's a shell issue. If you use variables in a shell script, they need to be quoted when using them, to avoid splitting arguments at spaces. In a case like this, it's cleanest use an array instead:

```bash
suffix=("-m" "not dangerous")
...
pytest "${suffix[@]}"
```

See [Quotes - Greg's Wiki](http://mywiki.wooledge.org/Quotes) and [Words... [Bash Hackers Wiki]](https://wiki.bash-hackers.org/syntax/words) for more explanation around it. 

I'd also recommend using [ShellCheck](https://www.shellcheck.net/), as it points out things like this:

```
[Line 3:]
suffix="-m 'not dangerous'"
        ^-- [SC2089](https://www.shellcheck.net/wiki/SC2089) (warning): Quotes/backslashes will be treated literally. Rewrite using set/"$@" or functions.
 
[Line 5:]
cd /usr/pytest/
^-- [SC2164](https://www.shellcheck.net/wiki/SC2164) (warning): Use 'cd ... || exit' or 'cd ... || return' in case cd fails.

Did you mean: ([apply this](javascript:applyFixIndex([1])), apply [all SC2164](javascript:applyFixCode(2164)))
cd /usr/pytest/ || exit
 
[Line 7:]
pytest ${suffix}
       ^-- [SC2090](https://www.shellcheck.net/wiki/SC2090) (warning): Quotes/backslashes in this variable will not be respected.
       ^-- [SC2086](https://www.shellcheck.net/wiki/SC2086) (info): Double quote to prevent globbing and word splitting.

Did you mean: ([apply this](javascript:applyFixIndex([3])), apply [all SC2086](javascript:applyFixCode(2086)))
pytest "${suffix}"
```

**Reply by lucky33newman** [reply]
Sir, yes sir! Salute! Actually, I am 100 per cent new to python and shell. So it is normal to make low level mistakes. X)
